addons/website_theme_install/models/ir_module_module.py

```diff
@@ -152,94 +125,151 @@ def _update_records(self, old, new):
+            ``theme_template_id`` will be set to NULL and the reference to the theme being removed
+            will be lost. However we do want the ophan to be deleted from the website when
+            we upgrade or delete the theme from the website.
+
+            ``website.page`` and ``website.menu`` don't have ``key`` field so we don't clean them.
+            TODO in master: add a field ``theme_id`` on the models to more cleanly compute orphans.
+        """
         self.ensure_one()
-        _logger.info('Unload theme %s for website %s from template.' % (self.mapped('name'), website.id))
-        menus = self._get_module_data('theme.website.menu')
-        menu_todel = menus.with_context(active_test=False).mapped('copy_ids').filtered(lambda m: m.website_id == website)
-        menu_todel.unlink()
-
-        pages = self._get_module_data('theme.website.page')
-        pages_todel = pages.with_context(active_test=False).mapped('copy_ids').filtered(lambda p: p.website_id == website)
-        pages_todel.unlink()
-
-        attachs = self._get_module_data('theme.ir.attachment')
-        attachs_todel = attachs.with_context(active_test=False).mapped('copy_ids').filtered(lambda a: a.website_id == website)
-        # double check - Some records that can be orphans or no more in the template view.
-        attachs_todel2 = self.env['ir.attachment'].with_context(active_test=False).search([('key', '=like', self.name + '_%'), ('website_id', '=', website.id)])
-        (attachs_todel + attachs_todel2).unlink()
-
-        views = self._get_module_data('theme.ir.ui.view')
-        views_todel = views.with_context(active_test=False).mapped('copy_ids').filtered(lambda v: v.website_id == website)
-        # double check - Some records that can be orphans or no more in the template view.
-        views_todel2 = self.env['ir.ui.view'].with_context(active_test=False).search([('key', '=like', self.name + '_%'), ('website_id', '=', website.id)])
-        (views_todel + views_todel2).unlink()
+
+        model = self.env[model_name]
+
+        if model_name in ('website.page', 'website.menu'):
+            return model
+
+        return model.with_context(active_test=False).search([
+            ('key', '=like', self.name + '.%'),
+            ('website_id', '=', website.id),
+            ('theme_template_id', '=', False),
+        ])
 
     @api.multi
-    def _remove_theme_on_website(self, website):
+    def _theme_cleanup(self, model_name, website):
+        """Remove orphan models of the current theme for the website ``website``."""
         self.ensure_one()
-        installed_deps = self + website.theme_id.upstream_dependencies(exclude_states=('',)).filtered(lambda x: x.name.startswith('theme_'))
-        for mod in installed_deps:
-            mod._unload_one_theme_module(website)
+        orphans = self._theme_find_orphans(model_name, website)
+        orphans.unlink()
+        return orphans
```

Comment:
unlink always return true.

I don't actually need to return anything, but the original method returned the created,updated,deleted models.

I don't know if I need to keep that feature or not, because those variables are then never used.

Reply:
or you return ids delete or nothing, but not true ;)

Reply:
Is it ok if _update_records also return nothing then?